pkg/providers/cloudstack/cloudstack.go

```diff
@@ -261,69 +261,49 @@ func (p *cloudstackProvider) MachineResourceType() string {
 		if err != nil {
 			return err
 		}
-	} else {
-		logger.Info("Provided control plane sshAuthorizedKey is not set or is empty, auto-generating new key pair...")
-		generatedKey, err := common.GenerateSSHAuthKey(p.writer)
+	}
+	if p.clusterConfig.Spec.ExternalEtcdConfiguration != nil {
+		_, err = p.processSshKeysForUsers(p.machineConfigs[p.clusterConfig.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name].Spec.Users, generatedKey)
 		if err != nil {
 			return err
 		}
-		controlPlaneUser.SshAuthorizedKeys[0] = generatedKey
-		useKeyGeneratedForControlplane = true
 	}
-	var workerUser v1alpha1.UserConfiguration
-	for _, workerNodeGroupConfiguration := range p.clusterConfig.Spec.WorkerNodeGroupConfigurations {
-		workerUser = p.machineConfigs[workerNodeGroupConfiguration.MachineGroupRef.Name].Spec.Users[0]
-		if len(workerUser.SshAuthorizedKeys[0]) > 0 {
-			workerUser.SshAuthorizedKeys[0], err = common.StripSshAuthorizedKeyComment(workerUser.SshAuthorizedKeys[0])
-			if err != nil {
-				return err
-			}
-		} else {
-			if useKeyGeneratedForControlplane { // use the same key
-				workerUser.SshAuthorizedKeys[0] = controlPlaneUser.SshAuthorizedKeys[0]
-			} else {
-				logger.Info("Provided worker sshAuthorizedKey is not set or is empty, auto-generating new key pair...")
-				generatedKey, err := common.GenerateSSHAuthKey(p.writer)
+	return nil
+}
+
+func (p *cloudstackProvider) processSshKeysForUsers(users []v1alpha1.UserConfiguration, generatedKey string) (string, error) {
+	var err error
+	for _, user := range users {
+		for i, key := range user.SshAuthorizedKeys {
+			if len(key) > 0 {
+				user.SshAuthorizedKeys[i], err = common.StripSshAuthorizedKeyComment(key)
 				if err != nil {
-					return err
+					return "", err
 				}
-				workerUser.SshAuthorizedKeys[0] = generatedKey
-				useKeyGeneratedForWorker = true
-			}
-		}
-	}
-	if p.clusterConfig.Spec.ExternalEtcdConfiguration != nil {
-		etcdUser := p.machineConfigs[p.clusterConfig.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name].Spec.Users[0]
-		if len(etcdUser.SshAuthorizedKeys[0]) > 0 {
-			etcdUser.SshAuthorizedKeys[0], err = common.StripSshAuthorizedKeyComment(etcdUser.SshAuthorizedKeys[0])
-			if err != nil {
-				return err
-			}
-		} else {
-			if useKeyGeneratedForControlplane { // use the same key as for controlplane
-				etcdUser.SshAuthorizedKeys[0] = controlPlaneUser.SshAuthorizedKeys[0]
-			} else if useKeyGeneratedForWorker {
-				etcdUser.SshAuthorizedKeys[0] = workerUser.SshAuthorizedKeys[0] // if cp key was provided by user, check if worker key was generated by cli and use that
 			} else {
-				logger.Info("Provided etcd sshAuthorizedKey is not set or is empty, auto-generating new key pair...")
-				generatedKey, err := common.GenerateSSHAuthKey(p.writer)
-				if err != nil {
-					return err
+				if len(generatedKey) == 0 {
+					logger.Info("Provided control plane sshAuthorizedKey is not set or is empty, auto-generating new key pair...")
```

Comment:
this log message shouldn't talk about control plane, should it? It could be any provided key, right?

Reply:
Good point!